dpcpp/test/solver/idr_kernels.cpp

```diff
@@ -0,0 +1,414 @@
+
+    gko::size_type k = 2;
+    gko::kernels::reference::idr::step_1(ref, nrhs, k, m.get(), f.get(),
+                                         r.get(), g.get(), c.get(), v.get(),
+                                         stop_status.get());
+    gko::kernels::dpcpp::idr::step_1(dpcpp, nrhs, k, d_m.get(), d_f.get(),
+                                     d_r.get(), d_g.get(), d_c.get(), d_v.get(),
+                                     d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(c, d_c, rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(v, d_v, rr<value_type>::value);
+}
+
+
+TEST_F(Idr, IdrStep2IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::size_type k = 2;
+    gko::kernels::reference::idr::step_2(ref, nrhs, k, omega.get(), v.get(),
+                                         c.get(), u.get(), stop_status.get());
+    gko::kernels::dpcpp::idr::step_2(dpcpp, nrhs, k, d_omega.get(), d_v.get(),
+                                     d_c.get(), d_u.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(u, d_u, rr<value_type>::value);
+}
+
+
+TEST_F(Idr, IdrStep3IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::size_type k = 2;
+    gko::kernels::reference::idr::step_3(
+        ref, nrhs, k, p.get(), g.get(), v.get(), u.get(), m.get(), f.get(),
+        alpha.get(), r.get(), x.get(), stop_status.get());
+    gko::kernels::dpcpp::idr::step_3(
+        dpcpp, nrhs, k, d_p.get(), d_g.get(), d_v.get(), d_u.get(), d_m.get(),
+        d_f.get(), d_alpha.get(), d_r.get(), d_x.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(g, d_g, 2 * rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(v, d_v, rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(u, d_u, rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(m, d_m, rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(f, d_f, 2 * rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(r, d_r, rr<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(x, d_x, rr<value_type>::value);
+}
+
+
+TEST_F(Idr, IdrComputeOmegaIsEquivalentToRef)
+{
+    initialize_data();
+
+    value_type kappa = 0.7;
+    gko::kernels::reference::idr::compute_omega(ref, nrhs, kappa, tht.get(),
+                                                residual_norm.get(),
+                                                omega.get(), stop_status.get());
+    gko::kernels::dpcpp::idr::compute_omega(dpcpp, nrhs, kappa, d_tht.get(),
+                                            d_residual_norm.get(),
+                                            d_omega.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(omega, d_omega, rr<value_type>::value);
+}
+
+
+TEST_F(Idr, IdrIterationOneRHSIsEquivalentToRef)
+{
+    int m = 123;
+    int n = 1;
+    auto ref_solver = ref_idr_factory->generate(mtx);
+    auto dpcpp_solver = dpcpp_idr_factory->generate(d_mtx);
+    auto b = gen_mtx(m, n);
+    auto x = gen_mtx(m, n);
+    auto d_b = Mtx::create(dpcpp);
+    auto d_x = Mtx::create(dpcpp);
+    d_b->copy_from(b.get());
+    d_x->copy_from(x.get());
```

Comment:
Minor: maybe we could make a function for this since it's used in many tests.

Reply:
I guess the matrix setup like set_up_apply_mtx()?